src/bin/poll_oneoff.rs

```diff
@@ -0,0 +1,205 @@
+use libc;
+use more_asserts::assert_gt;
+use std::{env, mem::MaybeUninit, process};
+use wasi::wasi_unstable;
+use wasi_misc_tests::{
+    open_scratch_directory,
+    utils::{cleanup_file, close_fd},
+    wasi_wrappers::wasi_path_open,
+};
+
+const CLOCK_ID: wasi_unstable::Userdata = 0x0123_45678;
+
+unsafe fn poll_oneoff_impl(
+    in_: &[wasi_unstable::Subscription],
+    nexpected: usize,
+) -> Vec<wasi_unstable::Event> {
+    let mut out: Vec<wasi_unstable::Event> = Vec::new();
+    out.resize_with(in_.len(), || {
+        MaybeUninit::<wasi_unstable::Event>::zeroed().assume_init()
+    });
+    let res = wasi_unstable::poll_oneoff(&in_, out.as_mut_slice());
+    assert!(res.is_ok(), "poll_oneoff should succeed");
+    assert_eq!(
+        res.unwrap(),
+        nexpected,
+        "poll_oneoff should return {} events",
+        nexpected
+    );
+    out
+}
+
+unsafe fn test_timeout() {
+    let clock = wasi_unstable::raw::__wasi_subscription_u_clock_t {
+        // NB The `identifier` field will become deprecated and removed from WASI.
+        // Ref: https://github.com/WebAssembly/WASI/pull/125
+        identifier: CLOCK_ID,
+        clock_id: wasi_unstable::CLOCK_MONOTONIC,
+        timeout: 5_000_000u64, // 5 milliseconds
+        precision: 0,
+        flags: 0,
+    };
+    let in_ = [wasi_unstable::Subscription {
+        userdata: CLOCK_ID,
+        type_: wasi_unstable::EVENTTYPE_CLOCK,
+        u: wasi_unstable::raw::__wasi_subscription_u { clock },
+    }];
+    let out = poll_oneoff_impl(&in_, 1);
+    let event = &out[0];
+    assert_eq!(
+        event.userdata, CLOCK_ID,
+        "the event.userdata should contain clock_id specified by the user"
+    );
+    assert_eq!(
+        event.error,
+        wasi_unstable::raw::__WASI_ESUCCESS,
+        "the event.error should be set to ESUCCESS"
+    );
+    assert_eq!(
+        event.type_,
+        wasi_unstable::EVENTTYPE_CLOCK,
+        "the event.type_ should equal clock"
+    );
+}
+
+unsafe fn test_stdin_read() {
```

Comment:
If the parent process closes `stdin`, it will not be inherited by the child process. As far as I understand, in such case, fd `1` should be closed and `poll` should return `POLLERR`.

Even if we don't support closed stdin, this needs a comment at the very least. We could probably try to write to stdin (if an error occurs, the stdin is definitely closed), try polling it (poll should return immediately), read the whole data from stdin, poll again (expecting a timeout)

We could also consider checking if the behavior for stdout and stderr is correct.

Reply:
This one is tricky. We don't actually store a raw host fd to the `stdin/out/err`. It is true though it is possible to remove a WASI fds pointing at the `stdin/out/err`, it will not close the raw host fds for the process running Wasm.

I'm trying to figure out if a case where a raw host fd to the `stdin` could be closed for the process running Wasm. Perhaps for now a comment would suffice? @sunfishcode thoughts?

Good idea about testing `stdout` and `stderr`. I'll see how difficult it will be to add it to the test case here.

Reply:
The shell script executing may have closed the `stdin` or `stdout`, for instance. I guess some CIs do it.

Reply:
This seems like something that the wasi-misc-tests framework isn't well equipped to handle. There are a variety of things we'd like to test for involving `stdin`/`stdout`/`stderr`, such as what if they're attached to special things like `/dev/random`, or fifos, or other things, that we can't do from within WASI itself. So while `wasi-misc-tests` is nice and simple, we may also need support for a test harness like Lucet's, which can set up a custom `WasiCtx`, run a WASI program with it, and test it from the outside.

Reply:
Hmm, that's a good point. How about we compromise a little bit here, and for the moment, test the positive route for `stdin/out/err` (NB I've already added a smoke test for the latter two), and in the future, when we figure out a better test harness, etc., we'll either augment this test case or create an additional one testing for ill conditions such as if the descriptors are attached to special things, fifos, etc.?

Reply:
Hmm, that's a good point. How about we compromise a little bit here, and for the moment, test the positive route for stdin/out/err (NB I've already added a smoke test for the latter two), and in the future, when we figure out a better test harness, etc., we'll either augment this test case or create an additional one testing for ill conditions such as if the descriptors are attached to special things, fifos, etc.?

Reply:
sounds okay, but please add a comment that we're assuming that stdin/stdout/stderr are open during the testsuite and state it in the README too

Reply:
Good point, will do!

Reply:
OK, I've now added the comment in the test case.